jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Server.java

```diff
@@ -151,6 +152,47 @@ public Server(@Name("threadPool") ThreadPool threadPool, @Name("scheduler") Sche
         addBean(FileSystemPool.INSTANCE, false);
     }
 
+    /**
+     * Get a new ComplianceViolation.Listener suitable for given Connector.
+     *
+     * @param connector the connector to base the ComplianceViolation.Listener off of.
+     * @return the ComplianceViolation.Listener implementation, or null if {@link HttpConnectionFactory#isRecordHttpComplianceViolations()} is false,
+     *   or there are no ComplianceViolation.Listener implementations registered.
+     */
+    public static ComplianceViolation.Listener newComplianceViolationListener(Connector connector)
+    {
+        HttpConnectionFactory httpConnectionFactory = connector.getConnectionFactory(HttpConnectionFactory.class);
+        if (httpConnectionFactory == null)
+            return null;
+
+        // Is this connector recording compliance violations?
+        if (!httpConnectionFactory.isRecordHttpComplianceViolations())
+            return null;
+
+        // Only add the ComplianceViolations instance if the recording of Compliance Violations is enabled
+        // This also means that any user provided ComplianceViolation.ListenerFactory beans will only be
+        // used when the configuration on the HttpConnectionFactory allows then to be used.
+
+        // Look for optional user provided ComplianceViolation.Listener and ComplianceViolation.ListenerFactory beans
+        List<ComplianceViolation.Listener> userListeners = new ArrayList<>();
+        for (org.eclipse.jetty.util.component.Container container: List.of(connector, connector.getServer()))
+        {
+            for (ComplianceViolation.Listener listener: container.getBeans(ComplianceViolation.Listener.class))
+                userListeners.add(listener);
+            for (ComplianceViolation.ListenerFactory listenerFactory: container.getBeans(ComplianceViolation.ListenerFactory.class))
+                userListeners.add(listenerFactory.newComplianceViolationListener());
+        }
+
+        // No listeners? then we are done
+        if (userListeners.isEmpty())
+            return null;
+        // Only 1 listener, just return it.
+        if (userListeners.size() == 1)
+            return userListeners.get(0);
+        // More than 1, establish ComplianceViolation.ListenerCollection for user listeners
+        return new ComplianceViolation.ListenerCollection(userListeners);
```

Comment:
Can some of this work be moved to onStart or done lazily on the first call.   It seams that if there are 2 listeners then a lot of work is done every single request.